First commit

talebzeghmi opened the pull request:
Seeking feedback on:
* Intake plugin pattern, does this plugin idiomatic Intake?
* How we open source and build.

talebzeghmi committed:
First commit

martindurant commented on intake_nested_yaml_catalog/nested_yaml_catalog.py:
```diff
@@ -0,0 +1,186 @@
+        super(YAMLCatalog, self).__init__(path=kwargs.get("name", ""), **kwargs)
+
+    def _load(self, reload=False):
+        self._dir = str(Path(__file__).resolve().parent)
```
This will be the directory containing the module, not the directory containing the YAML cat file.

martindurant commented on intake_nested_yaml_catalog/nested_yaml_catalog.py:
```diff
@@ -0,0 +1,186 @@
+        return dump(yaml_nested_cat, default_flow_style=False)
+
+
+global_registry['yaml_cat'] = YAMLCatalog
```
Please use new "entrypoints" API for registering https://intake.readthedocs.io/en/latest/making-plugins.html#entrypoints

martindurant commented on intake_nested_yaml_catalog/nested_yaml_catalog.py:
```diff
@@ -0,0 +1,186 @@
+        data = self._yaml(with_plugin=with_plugin)
+
+        # convert back to syntactic sugar of yaml_nested_cat
+        yaml_nested_cat = to_yaml_nested_cat(data, self.__module__ + "." + self.__class__.__name__)
```
`self.classname`

martindurant commented on intake_nested_yaml_catalog/nested_yaml_catalog.py:
```diff
@@ -0,0 +1,186 @@
+    version = __version__
+    container = 'catalog'
+    partition_access = None
+    name = 'yaml_cat'
```
Prefer something more specific

talebzeghmi replied:
renaming to `yaml_nested_cat_desugared`

martindurant commented on intake_nested_yaml_catalog/nested_yaml_catalog.py:
```diff
@@ -0,0 +1,186 @@
+
+# Override yaml_file_cat to support a catalog with taxonomy
+# yet fallback to YAMLFileCatalog if the file is in that format already
+global_registry['yaml_file_cat'] = YAMLFileNestedCatalog
```
I am not sure about overriding the existing drivers, it would probably be better to add new entries to prevent unexpected behaviour.

talebzeghmi replied:
How would the new plugin entry be the default for `intake.Catalog("cat.yaml")`?

martindurant replied:
It would not be, but would show up as `intake.open_hierarchical_yaml`, or `intake_zillow.open_catalog` (depending on the driver's name and how you choose to structure your repo). 

Maybe eventually it can take over, or there can be something in the top-level "metadata" key that marks the syntax, but I wouldn't like to swap out such a basic component yet, even if it only happens when this package is installed. I can imagine that swap leading to a lot of confusion.

talebzeghmi replied:
Instead of swapping the `yaml_file_cat` component, how about instead a PR to update `yaml_file_cat` to support nested hierarchies? 

That'd achieve two purposes:
1. Users of this plugin (and internal Zillow users) wouldn't be confused between `intake.open_yaml_cat()` and `intake.Catalog(path)` and `intake.open_yaml_nested_cat()`
2. Encourage the community to organize their datasets and avoid a data swamp.

talebzeghmi replied:
I see, I could provide user friendly functions via an intake-zillow (or intake-dal) plugin that exposes `intake_dal.Catalog(path), intake_dal.open_catalog()`.  What are your thoughts on providing this hierarchical YAML support within intake-core vs a plugin?

This intake-zillow/intake-dal plugin would provide helper functions for nested catalogs (and the dataset consistent access across storage system environments).

talebzeghmi replied:
Fixed!  Thank you for catching this hack!

I'm writing another plugin that can be invoked using `intake.DalCatalog("cat.yaml")` which would inherit from `YAMLFileNestedCatalog` and add the ability to abstracts a dataset over heterogenous storage systems.

Example entry:
```yaml
    my_datasource:
      driver: dal
      metadata:
        storage:
          default: 'local'
          local: 'csv://{{ CATALOG_DIR }}/data/user_events.csv'
          batch: 'parquet://{{ CATALOG_DIR }}/data/user_events.parquet'
```

`DalCatalog` plugin would do similar YAML transform but to:
```yaml
      my_datasource:
       driver:
         local:
           default: true
           class: intake.source.csv.CSVSource
           args:
             urlpath: '{{ CATALOG_DIR }}/data/user_events.csv'
         batch:
           class: intake_parquet.source.ParquetSource
           args:
             urlpath: '{{ CATALOG_DIR }}/data/user_events.parquet'
```

martindurant replied:
I think we should talk about this, there are many extension points here. The latter version, with multiple drivers, is already possible, and choices between arguments could already be done via user parameters or the alias class. I recommend you have a look at [that](https://github.com/intake/intake/blob/master/intake/source/base.py#L379) for some ideas of how I was thinking about this kind of problem before.

talebzeghmi replied:
`AliasSource` is very cool!  Wouldn't using `AliasSource` however require new entries in the Catalog possibly confusing users?

```yaml
    my_datasource:
      driver: intake.source.base.AliasSource
      args:
       target: "{{mode}}"
       mapping:
         local: my_datasource_local
         batch: my_datasource_batch
      parameters:
        mode:
          description: which to alias
          type: str
          default: local
          allowed: ["local", "batch"]
    my_datasource_local:
      driver: csv
      args:
        urlpath: '{{ CATALOG_DIR }}/data/user_events.csv'
    my_datasource_batch:
      driver: parquet
      args:
        urlpath: '{{ CATALOG_DIR }}/data/user_events.parquet'
```

talebzeghmi replied:
I think I see, you're suggesting to use the same pattern as `AliasSource` to front and specialize the source at access time!

martindurant replied:
Exactly so. There are, in fact, *many* patterns you could use, which is a benefit and perhaps drawback of an open system like Intake.

martindurant commented on intake_nested_yaml_catalog/tests/test_nested_yaml_catalog.py:
```diff
@@ -0,0 +1,52 @@
+    to_yaml_nested_cat  # noqa: F401
+
+
+class IntakeTest(TestCase):
```
Other intake repos prefer pytest and functions only, for the most part. Of course, author preference is respected here.

toumorokoshi replied:
+1 for pytest. It is also the default test framework in orbital-core:

https://github.com/zillow/orbital-core/blob/master/orbital_core/build/tasks.py#L18

https://docs.pytest.org/en/latest/

martindurant commented on intake_nested_yaml_catalog/tests/test_nested_yaml_catalog.py:
```diff
@@ -0,0 +1,52 @@
+
+class IntakeTest(TestCase):
+    def test_intake_yaml_cat(self):
+        catalog_path = str(Path(__file__).resolve().parent.joinpath(Path("yaml_cat.yaml")))
```
Perhaps just define the parent dir once for the whole module.

martindurant commented on README.rst:
```diff
@@ -0,0 +1,49 @@
+intake-nested-yaml-catalog
+==========================
+
+This is an Anaconda Intake plugin supporting a single YAML hierarchical
```
Anaconda doesn't own Intake, just because I was paid to write it :)  "Intake" alone will suffice here (with link to github or RTD)

martindurant commented on README.rst:
```diff
@@ -0,0 +1,49 @@
+Development
+--
+
+    ./uranium
```
Is it necessary? Probably needs some user and developer install instructions.

talebzeghmi replied:
I see the confusion. Will name that section to "Developer Instructions" and rename "Development" to "Build"

toumorokoshi replied:
you can only move things around the dev process and tools to a CONTRIBUTING.md file. Helps differentiate between what a consumer and a contributor need to know.

martindurant commented on README.rst:
```diff
@@ -0,0 +1,49 @@
+        args:
+          urlpath: s3://foo
+        driver: parquet
+      zestimate:
```
I haven't tried yet, but I wonder how you distinguish between sub-catalogs and entries, i.e., how come `entity.property` is a cat, `entity.property.zestimate` is an entry, but `entity.property.zestimate.args` is not.

talebzeghmi replied:
Thank you for bringing this up. to_yaml_cat_helper() transforms any YAML node that doesn't have a 'driver' key to an entry with drive='yaml_cat' (being renamed to yaml_nested_cat_desugared). In short, any YAML entry that doesn't have a 'driver' key is treated like a sub-catalog.

The pro is a clean YAML. The con is that it's not explicit in the YAML node path.

I like how the default Intake catalog specification explicitly declares that entries are

martindurant reviewed (commented):
Thanks, and nice to see your first commit here!

I have some general comments in the code, but my main one is whether it's really desirable to swap out the functionality of yaml_file_cat in Intake main; secondly, I think some more examples of usage may help motivate this somewhat (which are presumably loosely tied to Zillow's work-flow). Could there be a function to walk an existing hierarchical cat (made of files or in some other manner) and produce a hierarchical YAML view?

talebzeghmi commented:
It does seem like a hack to swap out `yaml_file_cat` plugin. 

The problem I'm trying to solve is two fold:
1. A nested catalog easy to use. `cat.entity.property.clean_attributes.read()`
2. Make it easy to declare a new hierarchy/namespace/node within the same single YAML file.

intake-nested-yaml-catalog syntactic sugar in a single YAML file:
```yaml
entity:
    description: "description1"
    property:
      clean_attributes:
        args:
          urlpath: s3://foo
        driver: parquet
      zestimate_project:
        zestimate:
          args:
            urlpath: '{{ CATALOG_DIR }}/zestimate.csv'
          description: zestimate description
          driver: csv
    user:
      user_profile:
        args:
          urlpath: s3://foo
        driver: parquet
```

The "desugared" approach requires multiple YAML files, and boilerplate (minimal) to reference them at each node of the hierarchy.

toumorokoshi commented on .gitignore:
```diff
@@ -0,0 +1,109 @@
+lib
+include
+man
+.Python
```
now that this project is public it can take advantage of uranium 2, which removes some of these annoying gitignores.

I'll send a PR over later.

toumorokoshi commented on intake_nested_yaml_catalog/nested_yaml_catalog.py:
```diff
@@ -0,0 +1,186 @@
+        if 'driver' in value and (value['driver'] == YAMLCatalog.name or value['driver'] == cls_name):
+            # unwind this entry
+            node = copy.deepcopy(value)
+            # node['driver'] = YAMLFileNestedCatalog.name
```
should this comment be removed?

toumorokoshi commented on setup.py:
```diff
@@ -0,0 +1,22 @@
+
+setup(
+    name="intake-nested-yaml-catalog",
+    setup_requires=["vcver", "setuptools-parcels"],
```
Reading up on versioneer, it produces it's own concept of version string. I'd suggest using one or the other exclusively, to avoid a mismatch inversions.

toumorokoshi commented on setup.py:
```diff
@@ -0,0 +1,22 @@
+    vcver={"path": BASE},
+    description="Intake single YAML hierarchical catalog.",
+    author="Zillow AI Platform",
+    author_email="aiplat@zillow.com",
```
is this intended to be a public-facing e-mail?

toumorokoshi commented on setup.py:
```diff
@@ -0,0 +1,22 @@
+        'pandas',
+        'PyYAML',
+        'deepmerge',
+        'orbital-core'
```
is there anything from orbital-core that is needed for the package to operate properly? I do see it is used in uranium's ubuild.py, but that's a dev time concern.

setup.py dependencies should enumerate the dependencies required to consume the package.

toumorokoshi commented on ubuild.py:
```diff
@@ -0,0 +1,11 @@
+current_build.config.set_defaults({
+    "package_name": "intake-nested-yaml-catalog",
+    "module": "intake_nested_yaml_catalog",
+    "binaries": {"enabled": True},
```
we don't need the binaries: enabled flag here. That's an internal Zillow construct used to get around some poorly packaged builds and a lack of support around wheels. Both of which can now be resolved, but we can't remove this flag internally until all package have been updated.

toumorokoshi commented on ubuild.py:
```diff
@@ -0,0 +1,11 @@
+    "binaries": {"enabled": True},
+})
+
+current_build.packages.install(".", develop=True)
```
this should be unneeded. orbital-core will install the develop package as part of the main task: https://github.com/zillow/orbital-core/blob/master/orbital_core/build/tasks.py#L6

toumorokoshi commented on intake_nested_yaml_catalog/tests/zestimate.csv:
```diff
@@ -0,0 +1,2 @@
+zpid,zestimate
```
as a suggestion I would choose non-internal zillow constructs. These seem benign but I would suggest making it a practice to scrub zillow-specific terminology and references.

toumorokoshi reviewed (approved):
Hope that helps!

talebzeghmi committed:
PR feedback

martindurant commented:
The following comment didn't seem to have sent. It may be a little obsolete now, but I'll send it anyway.

--

After talking with a few people, I think I remain in my original position, that it would be good to have the hierarchical-yaml implementation as a separate driver, at least for a while. Perhaps when development here stabilises, it could become one of the syntax formats accepted by open_catalog, or even replace yaml_file_cat sometime later.

To differentiate the syntax type, though, I would definitely include a metadata block at the top level, specifying the file's hierarchical nature. That will certainly help with adding a new path to open_catalog.

talebzeghmi commented:
> After talking with a few people, I think I remain in my original position, that it would be good to have the hierarchical-yaml implementation as a separate driver, at least for a while. Perhaps when development here stabilises, it could become one of the syntax formats accepted by open_catalog, or even replace yaml_file_cat sometime later.

makes sense, thank you!

> To differentiate the syntax type, though, I would definitely include a metadata block at the top level, specifying the file's hierarchical nature. That will certainly help with adding a new path to open_catalog.

@martindurant are you suggesting a required field under metadata to aid `open_catalog()`?  
Like:
```yaml
metadata:
  hierarchical: true
entity:
    description: "description"
    customer:
      customer_attributes:
        args:
          urlpath: s3://foo
        driver: parquet
```

martindurant commented:
>  are you suggesting a required field under metadata

Yes, that's what I meant.

talebzeghmi committed:
Assert hierarchical_catalog field in metadata

talebzeghmi committed:
Document hierarchical_catalog required field

talebzeghmi committed:
User friendly hierarchical_catalog required field assertion

talebzeghmi committed:
Create nested catalog w/ Intake types instead of YAML syntactic sugar
- Fixes several issues with parameter rendering
- Re-purposed intake Catalog tests for nested_yaml_catalog

talebzeghmi committed:
NestedCatalogEntry type and intake.Gui unit test

talebzeghmi committed:
Remove default __init__.py import of NestedYAMLFileCatalog

talebzeghmi committed:
Tab fix and description update

talebzeghmi committed:
Merge pull request #2 from zillow/tz/use-intake-catalog-types

Create nested catalog w/ Intake types instead of YAML syntactic sugar
